macros/src/main/scala-3/enumeratum/EnumMacros.scala

```diff
@@ -0,0 +1,212 @@
+  /** Finds any [A] in the current scope and returns an expression for a list of them. */
+  def findValuesImpl[A](using tpe: Type[A], q: Quotes): Expr[IndexedSeq[A]] = {
+    import q.reflect.*
+
+    val repr       = validateType[A]
+    val subclasses = enclosedSubClasses[A](q)(repr)
+
+    buildSeqExpr[A](q)(subclasses)
+  }
+
+  /** Given an A, provides its companion. */
+  def materializeEnumImpl[A, M: Type](using tpe: Type[A], q: Quotes): Expr[M] = {
+    import q.reflect.*
+
+    val repr = TypeRepr.of[A](using tpe)
+
+    if (!repr.typeSymbol.companionModule.exists) {
+      report.errorAndAbort(
+        s"""
+           |
+           |  Could not find the companion object for type ${repr.typeSymbol}.
+           |
+           |  If you're sure the companion object exists, you might be able to fix this error by annotating the
+           |  value you're trying to find the companion object for with a parent type (e.g. Light.Red: Light).
+           |
+           |  This error usually happens when trying to find the companion object of a hard-coded enum member, and
+           |  is caused by Scala inferring the type to be the member's singleton type (e.g. Light.Red.type instead of
+           |  Light).
+           |
+           |  To illustrate, given an enum:
+           |
+           |  sealed abstract class Light extends EnumEntry
+           |  case object Light extends Enum[Light] {
+           |    val values = findValues
+           |    case object Red   extends Light
+           |    case object Blue  extends Light
+           |    case object Green extends Light
+           |  }
+           |
+           |  and a method:
+           |
+           |  def indexOf[A <: EnumEntry: Enum](entry: A): Int = implicitly[Enum[A]].indexOf(entry)
+           |
+           |  Instead of calling like so: indexOf(Light.Red)
+           |                Call like so: indexOf(Light.Red: Light)
+         """.stripMargin
+      )
+    } else {
+      Ref(repr.typeSymbol.companionModule).asExprOf[M]
+    }
+  }
+
+  /** Makes sure that we can work with the given type as an enum:
+    *
+    * Aborts if the type is not sealed.
+    *
+    * @tparam T
+    *   the `Enum` type
+    */
+  private[enumeratum] def validateType[T](using q: Quotes, tpe: Type[T]): q.reflect.TypeRepr = {
+    import q.reflect.*
+
+    val repr = TypeRepr.of[T](using tpe)
+
+    if (!repr.classSymbol.exists(_.flags is Flags.Sealed)) {
+      report.errorAndAbort(
+        "You can only use findValues on sealed traits or classes"
+      )
+    }
+
+    repr
+  }
+
+  /** Returns a sequence of symbols for objects that implement the given type
+    *
+    * @tparam T
+    *   the `Enum` type
+    * @param tpr
+    *   the representation of type `T` (also specified by `tpe`)
+    */
+  private[enumeratum] def enclosedSubClasses[T](q: Quotes)(
+      tpr: q.reflect.TypeRepr
+  )(using tpe: Type[T]): List[q.reflect.TypeRepr] = {
+    import q.reflect.*
+
+    // TODO: Use SumOf?
```

Comment:
@cchantep can you please shed some light on this todo?

Reply:
Updated